include/jemalloc/internal/cache_bin.h

```diff
@@ -377,7 +405,74 @@ cache_bin_dalloc_easy(cache_bin_t *bin, void *ptr) {
 	return true;
 }
 
-/**
+/* Returns true if failed to stash (i.e. bin is full). */
+JEMALLOC_ALWAYS_INLINE bool
+cache_bin_stash(cache_bin_t *bin, void *ptr) {
+	if (cache_bin_full(bin)) {
+		return true;
+	}
+
+	/* Stash at the full position, in the [full, head) range. */
+	uint16_t low_bits_head = (uint16_t)(uintptr_t)bin->stack_head;
+	/* Wraparound handled as well. */
+	uint16_t diff = cache_bin_diff(bin, bin->low_bits_full, low_bits_head);
+	assert(!cache_bin_full(bin));
+	*(void **)((uintptr_t)bin->stack_head - diff) = ptr;
+
+	assert(!cache_bin_full(bin));
+	bin->low_bits_full += sizeof(void *);
+	cache_bin_assert_earlier(bin, bin->low_bits_full, low_bits_head);
+
+	return false;
+}
+
+JEMALLOC_ALWAYS_INLINE cache_bin_sz_t
+cache_bin_nstashed_get_internal(cache_bin_t *bin, cache_bin_info_t *info,
+    bool racy) {
+	cache_bin_sz_t ncached_max = cache_bin_info_ncached_max(info);
+	void **full = cache_bin_full_position_get(bin, info);
+
+	cache_bin_sz_t n = cache_bin_diff(bin, (uint16_t)(uintptr_t)full,
+	    bin->low_bits_full) / sizeof(void *);
```

Comment:
Added and stole your graph here :)

Reply:
@Lapenkov Sorry that I just realized that you were suggesting renaming the `full`, not the `low_bits_full` part. I somehow thought the `full` name was unchangeable. I like your idea better -- to clarify, we'd have `low_bound` (dervied), low_bits_full, stack_head, right?

Reply:
That works too. My main point was that it might seem that `low_bits_full` literally represents the low bits of `full`, which isn't the case. As long as these two concepts have different names, I'm ok with that.

Reply:
Agreed. Renamed the `full_position` parts to `low_bound`, and kept the low_bits_full (since we do compare against it to determine is_full_or_not anyway). Changes in https://github.com/jemalloc/jemalloc/pull/2173/commits/50fe35d24164175b1876591fb2467ddd6e6048c3

Other than that, squashed and rebased the feedback-addressing commit. No other changes planned. I'll merge if CI passes.